### PR TITLE
Find callchains

### PR DESCRIPTION
**Pull request by m5imunovic**
Add script for finding multiple paths between source (exact) and destination (regex) function.
Add relevant tests.
Add documentation.

### CODE REVIEW

File: safety-architecture/tools/callgraph-tool/doc/find_callchains_examples.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+chain is defined with second argument which is a regular expression matching
+one or more function names. The output is either the database of detected call
+chains in csv format or image of the detected subgraph
+
+optional arguments:
+  -h, --help            show this help message and exit
+  --out OUT             name of the output file containing detected chains
+  --direction {left,right,both}
+                        selects search direction.
+  --cutoff CUTOFF       select cutoff lenght for path search
```
Comment (henrirosten) [inline safety-architecture/tools/callgraph-tool/doc/find_callchains_examples.md:21]: cutoff length

File: safety-architecture/tools/callgraph-tool/doc/find_callchains_examples.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+```
+./find_callchains.py --from_function chain2 --to_function chain[0-9] --direction both --out chainX.jpg --calls calls.csv
+```
+This query results in a following image for a given database:
+
+<img src="chainX.jpg">
+<br /><br />
+
+## Larger example
+The examples given so far are intentionally contrived to illustrate the capabilites of the script. These results could have easily been read out from the image of the whole program graph. The usefulness of the script can be demonstrated on a more complicated codebase, such as that of a Linux kernel. Say we want to find out the receive system calls that use `sock_recvmsg` function. We can answer this question by issuing the following query:
```
Comment (henrirosten) [inline safety-architecture/tools/callgraph-tool/doc/find_callchains_examples.md:83]: capabilities

File: safety-architecture/tools/callgraph-tool/doc/find_callchains_examples.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+The script exports the results of a user search either to a `CSV` file (if `out` input parameter ends with .csv) or to an image file otherwise. The source function is always a literal match, while the destination function can be defined as regular expression. The search can be performed in both directions, i.e. both in
+callers of the source function - `left` search and callees of the source function - `right` search.
+
+# Usage examples
+
+
+Suppose that we have an example database for a program with a following call graph:
+
+<img src="all_callchains.jpg">
```
Comment (henrirosten) [inline safety-architecture/tools/callgraph-tool/doc/find_callchains_examples.md:45]: label on the 'printf' node in this image seems to have empty strings for the filename and line. Not a big deal, but the node label line seems strange when it contains only the colon ':'. Maybe use 'nan' for empty strings instead, or remove the colon if line and/or filename is empty.

File: safety-architecture/tools/callgraph-tool/doc/find_callchains_examples.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+## Search for callees
+
+We can also search in forward direction to find out the paths from a caller `main` to a specific function `say_hello` we are interested in:
+```
+./find_callchains.py --from_function main --to_function say_hello --direction right --out multi_chain.jpg --calls calls.csv
+```
+The command output is shown in:
+
+<img src="multi_chain.jpg">
```
Comment (henrirosten) [inline safety-architecture/tools/callgraph-tool/doc/find_callchains_examples.md:68]: Missing image: multi_chain.jpg

File: safety-architecture/tools/callgraph-tool/doc/find_callchains_examples.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+```
+This query results in a following image for a given database:
+
+<img src="chainX.jpg">
+<br /><br />
+
+## Larger example
+The examples given so far are intentionally contrived to illustrate the capabilites of the script. These results could have easily been read out from the image of the whole program graph. The usefulness of the script can be demonstrated on a more complicated codebase, such as that of a Linux kernel. Say we want to find out the receive system calls that use `sock_recvmsg` function. We can answer this question by issuing the following query:
+```
+./find_callshains --calls calls.csv, --from_function sock_recvmsg, --to_function ^__x64_sys_recv, --direction=left, --cutoff=12, --out=from_sock_recvmsg_to_x64_sys_recv_left.jpg
```
Comment (henrirosten) [inline safety-architecture/tools/callgraph-tool/doc/find_callchains_examples.md:85]: Are the commas in this command intentional?

Comment (m5imunovic) [reply safety-architecture/tools/callgraph-tool/doc/find_callchains_examples.md:85]: They remained as I copied command from IDE. Will be fixed

File: safety-architecture/tools/callgraph-tool/scripts/find_callchains.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+    required_named.add_argument("--from_function", help=help, required=True)
+    help = "function name where call chain ends (regex match)"
+    required_named.add_argument("--to_function", help=help, required=True)
+
+    help = "name of the output file containing detected chains"
+    parser.add_argument("--out", help=help, default="chains.csv")
+    choices = ["left", "right", "both"]
+    help = "selects search direction."
+    parser.add_argument("--direction", help=help, choices=choices, default="right")
+    help = "select cutoff lenght for path search"
```
Comment (henrirosten) [inline safety-architecture/tools/callgraph-tool/scripts/find_callchains.py:268]: cutoff *length*

Also, maybe the script output should somehow indicate - at least when the user does not specify a value for the cutoff argument - that the output results were generated with default cutoff of 10 and that there might actually be more matches if a larger cutoff was given

**Comment by henrirosten** [comment]
Overall, I think this is good, but please check the comments I added if you also think they are worth fixing, thanks.

File: safety-architecture/tools/callgraph-tool/doc/find_callchains_examples.md (hunk 3 above)
Comment (m5imunovic) [reply safety-architecture/tools/callgraph-tool/doc/find_callchains_examples.md:45]: Now showing NaN if value is missing

File: safety-architecture/tools/callgraph-tool/scripts/find_callchains.py (hunk 6 above)
Comment (m5imunovic) [reply safety-architecture/tools/callgraph-tool/scripts/find_callchains.py:268]: Added warning info when cutoff is activated. Moved default to None.

**Pull request merged**